modules/phase_field/src/ics/PolycrystalVoronoiVoidIC.C

```diff
@@ -0,0 +1,297 @@
+
+        if (std::abs(min_rij_1 - min_rij_2) > rij_diff_tol)
+          try_again = true;
+      }
+
+    } while (try_again == true);
+  }
+}
+
+Real
+PolycrystalVoronoiVoidIC::value(const Point & p)
+{
+  Real value = 0.0;
+
+  // Determine value for voids
+  Real void_value = MultiSmoothCircleIC::value(p);
+
+  // Determine value for grains
+  Real grain_value = grainValueCalc(p);
+
+  switch (_structure_type)
+  {
+    case 0:                 // assigning values for grains (order parameters)
+      if (grain_value == 0) // Not in this grain
+        value = grain_value;
+      else                             // in this grain, but might be in a void
+          if (void_value == _outvalue) // Not in a void
+        value = grain_value;
+      else if (void_value > _outvalue && void_value < _invalue) // On void interface
+        value = 1.0 - (void_value - _outvalue) / (_invalue - _outvalue);
+      else if (void_value == _invalue) // In a void, so op = 0
+        value = 0.0;
+      break;
+
+    case 1: // assigning values for voids (concentration)
+      value = void_value;
+      break;
+  }
+
+  return value;
+}
+
+RealGradient
+PolycrystalVoronoiVoidIC::gradient(const Point & p)
+{
+  RealGradient gradient;
+  RealGradient void_gradient = MultiSmoothCircleIC::gradient(p);
+
+  // Order parameter assignment assumes zero gradient (sharp interface)
+  switch (_structure_type)
+  {
+    case 1: // assigning gradient for voids
+      gradient = void_gradient;
+      break;
+  }
+
+  return gradient;
+}
+
+Real
+PolycrystalVoronoiVoidIC::grainValueCalc(const Point & p)
+{
+  Real val = 0.0;
+
+  unsigned int min_index =
+      PolycrystalICTools::assignPointToGrain(p, _centerpoints, _mesh, _var, _range.norm());
+
+  // If the current order parameter index (_op_index) is equal to the min_index,
+  // set the value to
+  // 1.0
+  if (_assigned_op[min_index] == _op_index)
+    val = 1.0;
+
+  if (val > 1.0)
+    val = 1.0;
+
+  if (val < 0.0)
+    val = 0.0;
+
+  return val;
+}
+
+void
+PolycrystalVoronoiVoidIC::computeGrainCenters()
+{
+  if (_op_num > _grain_num)
+    mooseError("ERROR in PolycrystalVoronoiVoidIC: Number of order parameters "
+               "(op_num) can't be "
+               "larger than the number of grains (grain_num)");
+
+  // Initialize vectors
+  _centerpoints.resize(_grain_num);
+  _assigned_op.resize(_grain_num);
+
+  // Randomly generate the centers of the individual grains represented by the
+  // Voronoi tessellation
+  for (unsigned int grain = 0; grain < _grain_num; grain++)
+  {
+    for (unsigned int i = 0; i < LIBMESH_DIM; i++)
+      _centerpoints[grain](i) = _bottom_left(i) + _range(i) * MooseRandom::rand();
+
+    if (_columnar_3D)
+      _centerpoints[grain](2) = _bottom_left(2) + _range(2) * 0.5;
+  }
+
+  // Assign grains to specific order parameters in a way that maximizes the
+  // distance
+  _assigned_op = PolycrystalICTools::assignPointsToVariables(_centerpoints, _op_num, _mesh, _var);
```

Comment:
To be honest I'd like to try to get it in to Moose as soon as we can... this is being moved from MARMOT to MOOSE, and the MR to remove it from MARMOT has already been merged. What is driving this is we have been trying to get these objects accessible to an Argonne collaborator who's currently unable to access MARMOT because he is a foreign national and they are not allowing non-INL FN's to get access to INL HPC right now, which he would need for Gitlab. (And apparently as of last week, even INL FN's are on hold for accessing Marmot.) Our collaborator needs these methods to do some simulations of thermal conductivity for the ATF HIP that is supposed to be done this fiscal year, hence the time sensitivity.

Could we merge this as-is for now, and fix it up when your changes to the OP assignment algorithms are done? Since it's already in Marmot like this, it's going to need to be re-worked at some point, whether it's in Marmot or Moose.